include/libsurvive/survive_api.h

```diff
@@ -0,0 +1,63 @@
+#include "survive_types.h"
+#include <stdbool.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+	struct SurviveAsyncContext;
+	typedef struct SurviveAsyncContext SurviveAsyncContext;
+
+	/***
+	 * Initialize a new instance of an async context -- mirrors survive_init
+	 * @return Pointer to the async context
+	 */
+	SURVIVE_EXPORT SurviveAsyncContext *survive_async_init(int argc, char *const *argv);
+
+	/**
+	 * Close all devices and free all memory associated with the given context
+	 */
+	SURVIVE_EXPORT void survive_async_close(SurviveAsyncContext* actx);
+
+	/**
+	 * Start the background thread which processes images.
+	 */
+	SURVIVE_EXPORT void survive_async_start_thread(SurviveAsyncContext* actx);
+
+	/**
+	 * @return true iff the background thread is still running
+	 */
+	SURVIVE_EXPORT bool survive_async_is_running(SurviveAsyncContext *actx);
+
+	struct SurviveAsyncObject;
+	typedef struct SurviveAsyncObject SurviveAsyncObject;
+
+	/**
+	 * Get the first known object. Note that the first n objects are the lighthouses; for however many you are
+	 * configured for.
+	 */
+	SURVIVE_EXPORT const SurviveAsyncObject* survive_async_get_first_tracked(SurviveAsyncContext* actx);
+	/**
+	 * Get the next known object from a current one.
+	 */
+	SURVIVE_EXPORT const SurviveAsyncObject* survive_async_get_next_tracked(SurviveAsyncContext* actx, const SurviveAsyncObject* curr);
+
+	/**
+	 * Gets the next object which has been updated since we last looked at it with this function
+	 */
+	SURVIVE_EXPORT const SurviveAsyncObject *survive_async_get_next_updated(SurviveAsyncContext *actx);
```

Comment:
I don't like that this function keeps internal state about the caller, and therefore can only ever be used by one caller (and if an application has more than one caller, it fails by mysteriously dropping events for each separate caller).  If we're going to do an synchronous "get next message" then I would propose we do it right so that any number of callers can use it without  stomping on each other.  Basically, make the function complete enough that you could reasonably use it entirely instead of async notifications.  

I'd actually set a few requirements on this function: 
1. Support an event depth of greater than 1 (preferably up to the number of objects we have).  
2. Consumable by any number of different callers, without stomping on each other
3. Supports a timed wait and timeout when waiting for the "next" event that hasn't happened yet.

Something like: 
void survive_async_get_next_event(IN unsigned int previousEventId ,OUT unsigned int *newEventId, OUT bool *skipped, OUT SurviveAsyncObject *sao, IN SurviveAsyncContext *actx, IN int timoutInMs);

To get started, call with a previousEvent of 0 or -1 (whatver magic number we choose).  Then, keep calling repeatedly, passing in the previously returned newEventId into the previousEventId parameter.  The "skipped" parameter returns false when you've lagged to far behind and events were lost.  The "sao" parameter is the handle  to the survive object whose pose changed.  The timeout is key here, too.  If there's a next event immediately available, it would return right away always.  If there is not a next event, then the function would wait up to the timeout value for an event to be available before returning.  If you timeout, the returned newEventId would equal the previousEventId passed in.  You could also have a bool.  

At some point in the future, we might also want to extend the types of events you could receive to more than just poses.  You could also receive button press events, for example.  Not sure how far down that path to go...

Reply:
To be clear on what the purpose of this function is: It isn't to grab the 'next' position; and it isn't a sync call -- it will never hard block. 

What it does instead is tell you the last position that it saw with the timestamp of that position. 

It currently does support an event depth of exactly the number of objects you have; see implementation. 

I'm against the idea of timed wait / timeout based calls; it'd be a API constraint we'd have to live with and I'm not sure the use case.

Maybe something like:

SurviveAsyncObject * survive_async_get_event_after(IN const SurviveAsyncContext *actx, INOUT uint64_t* since);

Ideally we make uint64_t an absolute timestamp in us that is consistent across all objects. There are details there to be worked on but I think its a reasonable direction.

A somewhat important note about returning the pointer vs 'OUT SurviveAsyncObject *sao': It gives us a very wide latitude in implementation if we don't expose the actual struct layout and forces users to use proper accessors. If we had something like 'OUT SurviveAsyncObject *sao' as a parameter, SurviveAsyncObject couldn't be an opaque type, thus it has to be either SurviveAsyncObject **sao or a return pointer as seen above.